cli/lib/gateway.js

```diff
@@ -1,104 +1,200 @@
+          cluster.terminate(() => {
+            console.log('Stop completed');
+            socket.sendMessage(true);
+            process.exit(0);
+          });
+        } else if (message.command == 'status') {
+          var activeWorkers = cluster.activeWorkers();
+          socket.sendMessage(activeWorkers ? activeWorkers.length : 0);
+        }
+      });
+    });
+    var opt = {};
+    opt.args = [JSON.stringify(args)];
+    opt.timeout = 10;
+
+    var cluster = reloadCluster(path.join(__dirname, 'start-agent.js'), opt);
+
+    cluster.run();
+    console.log('PROCESS PID : '+ process.pid);
+
+    process.on('exit', () => {
+      if (!isWin) {
+        console.log('Removing the socket file as part of cleanup');
+        fs.unlinkSync(ipcPath);
       }
+    });
+
+    process.on('SIGTERM', () => {
+      process.exit(0);
+    });
+
+    process.on('SIGINT', () => {
+      process.exit(0);
+    });
 
+  });
+};
+
+Gateway.prototype.reload = (options) => {
+  const source = configLocations.getSourcePath(options.org, options.env);
+  const cache = configLocations.getCachePath(options.org, options.env);
+  const keys = {key: options.key, secret: options.secret};
+
+  var socket = new JsonSocket(new net.Socket()); //Decorate a standard net.Socket with JsonSocket
+  socket.on('connect', () => {
+    edgeconfig.get({source: source, keys: keys}, (err, config) => {
       if (err) {
         const exists = fs.existsSync(cache);
         console.error("failed to retieve config from gateway. continuing, will try cached copy..");
         console.error(err);
-        if(!exists){
-          return cb('cache configuration '+cache+' does not exist. exiting.');
-        }else{
-          console.log('using cached configuration from %s',cache);
-          config = edgeconfig.load({source:cache})
-          if(options.port){
-            config.edgemicro.port = parseInt(options.port);
-          }
+        if (!exists) {
+          console.error('cache configuration ' + cache + ' does not exist. exiting.');
+          return;
+        } else {
+          console.log('using cached configuration from %s', cache);
+          config = edgeconfig.load({source: cache})
         }
       } else {
         edgeconfig.save(config, cache);
       }
 
-      if (options.cluster) {
-
-        const numWorkers = Number(options.processes || require('os').cpus().length);
-        cluster.setupMaster();
-        const argv = cluster.settings ? cluster.settings.execArgv || [] : [];
-        var j = 0;
-        argv && argv.forEach((arg) => {
-          if (arg.includes('--debug-brk=')) {
-            argv[j] = arg.replace('--debug-brk', '--debug')
-          }
-          j++;
-        })
-        cluster.isMaster && console.log("starting in cluster mode: number workers: " + numWorkers)
-        // Fork workers.
-        for (var i = 0; i < numWorkers; i++) {
-          cluster.fork();
-        }
+      gateway(config);
```

Comment:
same as above, please remove (even though it was already present, may as well take the chance to clean it up)